### PR TITLE
release gil more

### PR DESCRIPTION
**Pull request by jbrockmendel**
cc @mroeschke since a lot of this touches libwindow

### CODE REVIEW

**Comment by mroeschke** [comment]
Thanks @jbrockmendel. I am doing some heavy refactoring in `rolling.py` and `window.pyx` right now, but it looks like your changes shouldn't conflict heavily with what I'm changing.

File: pandas/_libs/window.pyx [hunk 1]
```diff
@@ -1679,18 +1677,18 @@ def roll_generic(object obj,
...
-def roll_weighted_mean(float64_t[:] values, float64_t[:] weights,
-                       int minp):
+def roll_weighted_mean(float64_t[:] values, float64_t[:] weights, int minp):
+    # `.base` to access underlying ndarray
     return _roll_weighted_sum_mean(values, weights, minp, avg=1)
 
 
-def _roll_weighted_sum_mean(float64_t[:] values, float64_t[:] weights,
-                            int minp, bint avg):
+cdef _roll_weighted_sum_mean(float64_t[:] values, float64_t[:] weights,
```
Comment (WillAyd) [inline pandas/_libs/window.pyx]: `cdef void` (unless this returns something)

File: pandas/_libs/window.pyx [hunk 2]
```diff
@@ -1797,29 +1796,30 @@ def ewma(float64_t[:] vals, float64_t com,
...
-                        old_wt += new_wt
-                    else:
-                        old_wt = 1.
-        elif is_observation:
-            weighted_avg = cur
+    with nogil:
+        for i in range(1, N):
+            cur = vals[i]
+            is_observation = (cur == cur)
+            nobs += <int64_t>(1 if is_observation else 0)
```
Comment (WillAyd) [inline pandas/_libs/window.pyx]: Is the `if...else` really required here? Wouldn't cast to int64 handle that?

File: pandas/_libs/algos_common_helper.pxi.in [hunk 3]
```diff
@@ -38,39 +38,41 @@ def diff_2d(ndarray[diff_t, ndim=2] arr,
         #  warnings about unreachable code
         sx, sy = (<object>arr).shape
         if arr.flags.f_contiguous:
-            if axis == 0:
-                if periods >= 0:
-                    start, stop = periods, sx
+            with nogil:
```
Comment (WillAyd) [inline pandas/_libs/algos_common_helper.pxi.in]: Can the `nogil` here not just precede the `if...else`?

**Review by WillAyd — CHANGES_REQUESTED** [review]
Sure makes sense. Minor things

File: pandas/_libs/window.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/window.pyx]: sure.  does this affect perf or is it just more explicit?

File: pandas/_libs/algos_common_helper.pxi.in (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_common_helper.pxi.in]: That'd be nice, but no.  The `arr.flags.f_contiguous` check is in python-space

File: pandas/_libs/window.pyx (hunk 1 above)
Comment (WillAyd) [reply pandas/_libs/window.pyx]: I'm coming from the explicit angle but I can't imagine it would hurt either

File: pandas/_libs/algos_common_helper.pxi.in (hunk 3 above)
Comment (WillAyd) [reply pandas/_libs/algos_common_helper.pxi.in]: Can you assign to a bool before the block and check that at the top instead of staying in the Python space? On first impression I was expecting the difference branches to have different GIL considerations, hence the ask

File: pandas/_libs/window.pyx (hunk 2 above)
Comment (jbrockmendel) [reply pandas/_libs/window.pyx]: Huh i thought this was necessary, but now looks like it isnt.  will change

File: pandas/_libs/window.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/window.pyx]: actually looks like this should be `cdef ndarray[float64_t]`, will update

File: pandas/_libs/algos_common_helper.pxi.in (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_common_helper.pxi.in]: good idea, will do

File: pandas/_libs/window.pyx [hunk 4]
```diff
@@ -1702,64 +1699,65 @@ def _roll_weighted_sum_mean(float64_t[:] values, float64_t[:] weights,
...
-                continue
+        else:
+            for win_i in range(win_n):
+                val_win = weights[win_i]
+                if val_win != val_win:
+                    continue
 
-            for in_i from 0 <= in_i < in_n - (win_n - win_i) + 1:
-                val_in = values[in_i]
+                for in_i in range(in_n - (win_n - win_i) + 1):
```
Comment (WillAyd) [inline pandas/_libs/window.pyx]: Is this totally equivalent? I'm not familiar with the for...from syntax but seems like the old one *could* short circuit if `in_i` falls in between 0 and the upper bound of the range here

File: pandas/_libs/window.pyx [hunk 5]
```diff
@@ -1797,29 +1796,30 @@ def ewma(float64_t[:] vals, float64_t com,
...
-                        old_wt += new_wt
-                    else:
-                        old_wt = 1.
-        elif is_observation:
-            weighted_avg = cur
+    with nogil:
+        for i in range(1, N):
+            cur = vals[i]
+            is_observation = (cur == cur)
+            nobs += <int64_t>is_observation
```
Comment (WillAyd) [inline pandas/_libs/window.pyx]: I think this cast is unnecessary; at least C should widen the bool type to int64 as part of the addition

File: pandas/_libs/window.pyx [hunk 6]
```diff
@@ -1882,55 +1884,57 @@ def ewmcov(float64_t[:] input_x, float64_t[:] input_y,
...
-            if not bias:
-                numerator = sum_wt * sum_wt
-                denominator = numerator - sum_wt2
-                if (denominator > 0.):
-                    output[i] = ((numerator / denominator) * cov)
+        for i in range(1, N):
+            cur_x = input_x[i]
+            cur_y = input_y[i]
+            is_observation = ((cur_x == cur_x) and (cur_y == cur_y))
+            nobs += <int64_t>is_observation
```
Comment (WillAyd) [inline pandas/_libs/window.pyx]: Same comment on cast

**Review by WillAyd — COMMENTED** [review]
looks good to me. Minor things to consider otherwise happy with merge

File: pandas/_libs/window.pyx (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/window.pyx]: > if in_i falls in between 0 and the upper bound of the range here

in_i is the loop variable.  I don't understand the failure mode you're asking about

File: pandas/_libs/window.pyx [hunk 7]
```diff
@@ -1702,64 +1699,65 @@ def _roll_weighted_sum_mean(float64_t[:] values, float64_t[:] weights,
...
-            if c < minp:
-                output[in_i] = NaN
+            for in_i in range(in_n):
+                c = counts[in_i]
+                if c < minp:
+                    output[in_i] = NaN
 
-    return output
+    # `.base` to access underlying ndarray
+    return output.base
```
Comment (jreback) [inline pandas/_libs/window.pyx]: why is this needed?

**Comment by jreback** [comment]
does this show any perf benefits on roll ing sum?

File: pandas/_libs/window.pyx (hunk 7 above)
Comment (jbrockmendel) [reply pandas/_libs/window.pyx]: at this point `output` is a memoryview and we need an ndarray.  this is how you do that without creating a new object

Comment (jreback) [reply pandas/_libs/window.pyx]: IIRC don't we ucommonly use ``np.asarray(output)`` ? I think this is more explict (and the same perf wise).

**Comment by jreback** [comment]
pls rebase

File: pandas/_libs/window.pyx (hunk 7 above)
Comment (jbrockmendel) [reply pandas/_libs/window.pyx]: we're not entirely consistent about it, but i agree on the explicit thing.  On the margin it must be slightly less performant because it calls into python-space, but that'll be in the "too small to measure" category

Comment (jreback) [reply pandas/_libs/window.pyx]: yeah i like it because its more explicit.

**Comment by jbrockmendel** [comment]
rebased, changed output.base to np.asarray(output), will run some asvs

**Comment by jbrockmendel** [comment]
Wow, way better than I had expected:

```
$ asv continuous -E virtualenv -f 1.1 master HEAD -b rolling
[...]
       before           after         ratio
     [023fa0cf]       [773c9f94]
     <master>         <lessgil2>
-        48.6±4ms       43.6±0.8ms     0.90  rolling.Quantile.time_quantile('DataFrame', 10, 'float', 0.5, 'linear')
-        69.6±5ms       61.8±0.9ms     0.89  rolling.Quantile.time_quantile('DataFrame', 1000, 'float', 0.5, 'linear')
-        49.4±6ms       42.6±0.6ms     0.86  rolling.Methods.time_rolling('DataFrame', 10, 'float', 'median')
-      2.05±0.2ms      1.71±0.05ms     0.83  rolling.Methods.time_rolling('DataFrame', 1000, 'float', 'std')
-      4.78±0.3ms       3.91±0.2ms     0.82  rolling.VariableWindowMethods.time_rolling('DataFrame', '1h', 'float', 'max')
-        78.9±8ms       62.6±0.7ms     0.79  rolling.Quantile.time_quantile('DataFrame', 1000, 'float', 0.5, 'midpoint')
-        54.0±4ms       42.8±0.9ms     0.79  rolling.Quantile.time_quantile('Series', 10, 'int', 0.5, 'linear')
-      3.25±0.2ms      2.54±0.09ms     0.78  rolling.VariableWindowMethods.time_rolling('Series', '1h', 'int', 'skew')
-      4.74±0.3ms       3.64±0.2ms     0.77  rolling.VariableWindowMethods.time_rolling('DataFrame', '1h', 'int', 'max')
-      2.21±0.2ms      1.69±0.02ms     0.76  rolling.ExpandingMethods.time_expanding('DataFrame', 'float', 'std')
-     2.39±0.05ms      1.79±0.06ms     0.75  rolling.VariableWindowMethods.time_rolling('DataFrame', '1d', 'float', 'count')
-     2.49±0.02ms      1.85±0.07ms     0.75  rolling.Quantile.time_quantile('DataFrame', 1000, 'float', 1, 'higher')
-      4.18±0.2ms      3.10±0.09ms     0.74  rolling.VariableWindowMethods.time_rolling('DataFrame', '1h', 'int', 'std')
-      1.64±0.3ms      1.21±0.02ms     0.74  rolling.Quantile.time_quantile('DataFrame', 10, 'int', 0, 'higher')
-      3.60±0.2ms      2.65±0.07ms     0.74  rolling.Methods.time_rolling('Series', 1000, 'int', 'count')
-      4.82±0.1ms       3.53±0.1ms     0.73  rolling.VariableWindowMethods.time_rolling('DataFrame', '50s', 'float', 'min')
-      2.82±0.1ms      2.05±0.06ms     0.73  rolling.VariableWindowMethods.time_rolling('Series', '50s', 'int', 'sum')
-      2.46±0.2ms      1.77±0.05ms     0.72  rolling.ExpandingMethods.time_expanding('DataFrame', 'float', 'min')
-      5.46±0.3ms       3.91±0.2ms     0.71  rolling.VariableWindowMethods.time_rolling('Series', '50s', 'float', 'min')
-      2.79±0.1ms      1.96±0.04ms     0.70  rolling.Quantile.time_quantile('Series', 1000, 'float', 0, 'nearest')
-        19.1±2ms       13.2±0.2ms     0.69  rolling.Pairwise.time_pairwise(None, 'cov', True)
-     2.48±0.03ms      1.71±0.04ms     0.69  rolling.Quantile.time_quantile('DataFrame', 1000, 'float', 0, 'lower')
-     2.91±0.09ms      1.98±0.03ms     0.68  rolling.Quantile.time_quantile('Series', 1000, 'float', 0, 'midpoint')
-      3.83±0.2ms      2.60±0.07ms     0.68  rolling.Methods.time_rolling('Series', 10, 'float', 'count')
-      5.41±0.2ms      3.67±0.04ms     0.68  rolling.VariableWindowMethods.time_rolling('Series', '50s', 'int', 'max')
-      2.80±0.2ms      1.83±0.04ms     0.65  rolling.Quantile.time_quantile('Series', 10, 'float', 1, 'nearest')
-      3.85±0.2ms       2.51±0.2ms     0.65  rolling.VariableWindowMethods.time_rolling('DataFrame', '1h', 'int', 'skew')
-      2.71±0.1ms      1.74±0.04ms     0.64  rolling.Quantile.time_quantile('DataFrame', 1000, 'float', 1, 'lower')
-      5.99±0.2ms       2.19±0.4ms     0.37  rolling.EWMMethods.time_ewm('Series', 1000, 'int', 'mean')
-     5.69±0.09ms       2.07±0.4ms     0.36  rolling.EWMMethods.time_ewm('Series', 10, 'int', 'mean')
-      5.78±0.2ms      1.65±0.07ms     0.28  rolling.EWMMethods.time_ewm('Series', 10, 'float', 'mean')
-      5.90±0.6ms      1.67±0.03ms     0.28  rolling.EWMMethods.time_ewm('Series', 1000, 'float', 'mean')
-      5.49±0.1ms       1.42±0.2ms     0.26  rolling.EWMMethods.time_ewm('DataFrame', 1000, 'int', 'mean')
-      5.46±0.2ms      1.39±0.07ms     0.25  rolling.EWMMethods.time_ewm('DataFrame', 10, 'float', 'mean')
-      5.61±0.2ms      1.43±0.04ms     0.25  rolling.EWMMethods.time_ewm('DataFrame', 10, 'int', 'mean')
-      5.43±0.2ms      1.31±0.02ms     0.24  rolling.EWMMethods.time_ewm('DataFrame', 1000, 'float', 'mean')
-      13.2±0.2ms       2.87±0.5ms     0.22  rolling.EWMMethods.time_ewm('DataFrame', 10, 'int', 'std')
-      13.4±0.1ms       2.42±0.4ms     0.18  rolling.EWMMethods.time_ewm('Series', 10, 'int', 'std')
-      13.3±0.2ms      2.36±0.08ms     0.18  rolling.EWMMethods.time_ewm('DataFrame', 1000, 'int', 'std')
-      13.3±0.4ms      2.31±0.05ms     0.17  rolling.EWMMethods.time_ewm('DataFrame', 10, 'float', 'std')
-      13.4±0.5ms      2.28±0.03ms     0.17  rolling.EWMMethods.time_ewm('DataFrame', 1000, 'float', 'std')
-      13.2±0.3ms      2.22±0.05ms     0.17  rolling.EWMMethods.time_ewm('Series', 10, 'float', 'std')
-        15.7±2ms       2.32±0.1ms     0.15  rolling.EWMMethods.time_ewm('Series', 1000, 'int', 'std')
-        15.5±2ms       2.15±0.1ms     0.14  rolling.EWMMethods.time_ewm('Series', 1000, 'float', 'std')
```

File: pandas/_libs/window.pyx (hunk 4 above)
Comment (WillAyd) [reply pandas/_libs/window.pyx]: I think just my misunderstanding of the invariant. Haven't come across the previous syntax

**Review by WillAyd — APPROVED** [review]
lgtm. Whatsnew note for perf boost might be nice, especially for the expanding benchmarks. Otherwise over to @jreback

**Comment by jreback** [comment]
lgtm. nice perf boosts! what's up with the windows build?

**Review by jreback — APPROVED** [review]

**Comment by jbrockmendel** [comment]
> what's up with the windows build?

Looks unrelated:

```
Windows fatal exception: access violation
[...]
worker 'gw1' crashed while running 'pandas/tests/io/json/test_compression.py::test_to_json_compression[gzip-True-False]'
```

**Comment by jbrockmendel** [comment]
green

**Comment by jreback** [comment]
thanks, very nice
